File: src/main/java/kata/MeetupEvent.java
package kata;

import java.time.LocalDateTime;
import java.util.List;

import static java.util.Comparator.comparing;

public class MeetupEvent {

    private final Long id;
    private final int capacity;
    private final String eventName;
    private final LocalDateTime startTime;
    private final List<Subscription> subscriptions;

    public MeetupEvent(Long id, int capacity, String eventName, LocalDateTime startTime, List<Subscription> subscriptions) {
        this.id = id;
        this.capacity = capacity;
        this.eventName = eventName;
        this.startTime = startTime;
        this.subscriptions = subscriptions;
    }

    public Long getId() {
        return id;
    }

    public int getCapacity() {
        return capacity;
    }

    public String getEventName() {
        return eventName;
    }

    public LocalDateTime getStartTime() {
        return startTime;
    }

    public List<Subscription> getSubscriptions() {
        return subscriptions == null ? List.of() : subscriptions;
    }

    public MeetupEvent withCapacity(int capacity) {
        return new MeetupEvent(id, capacity, eventName, startTime, subscriptions);
    }

    public Subscription getSubscription(String userId) {
        return getSubscriptions().stream()
                .filter(it -> it.getUserId().equals(userId))
                .findAny()
                .orElse(null);
    }

    public boolean isInWaitingList(String userId) {
        return getSubscription(userId).isInWaitingList();
    }

    public List<Subscription> getWaitingList() {
        return getSubscriptions().stream()
                .filter(it -> it.isInWaitingList())
                .toList();
    }

    public List<Subscription> getParticipants() {
        return getSubscriptions().stream()
                .filter(it -> it.isInWaitingList() == false)
                .sorted(comparing(Subscription::getRegistrationTime))
                .toList();
    }

    public void remove(String userId) {
        getSubscriptions().stream()
                .filter(it -> it.getUserId().equals(userId))
                .findFirst()
                .ifPresent(it -> subscriptions.remove(it));
    }

    public List<String> getUsers() {
        return getSubscriptions().stream()
                .map(Subscription::getUserId)
                .toList();
    }

    public void changeFromWaitingListToParticipant(String userId) {
        getSubscriptions().stream()
                .filter(it -> it.getUserId().equals(userId))
                .findFirst()
                .ifPresent(it -> it.confirm());
    }
}

File: src/main/kotlin/kata/MeetupEvent.kt
package kata

import java.time.Instant
import java.time.LocalDateTime
import java.util.*
import java.util.stream.Stream


data class MeetupEventState(
  val id: Long,
  val capacity: Int,
  val eventName: String,
  val startTime: LocalDateTime,
  val subscriptions: Subscriptions = Subscriptions(),
  val lastAppliedEventIndex: Int = -1,
) {

  val waitingList: List<Subscription>
    get() = subscriptions.waitingList

  val participants: List<Subscription>
    get() = subscriptions.participants

  val isFull: Boolean
    get() = participants.size == capacity

  fun hasSubscriptionFor(userId: String): Boolean {
    return subscriptions.findBy(userId) != null
  }

  fun applyEvent(event: MeetupBaseEvent): MeetupEventState {
    val newState = when (event) {
      is MeetupEventRegistered -> MeetupEventState(event.id, event.eventCapacity, event.eventName, event.startTime)
      is MeetupEventCapacityIncreased -> apply(event)
      is UserSubscribedToMeetupEvent -> apply(event)
      is UserAddedToMeetupEventWaitingList -> apply(event)
      is UserCancelledMeetupSubscription -> apply(event)
      is UserMovedFromWaitingListToParticipants -> apply(event)
      is UsersMovedFromWaitingListToParticipants -> apply(event)
    }
    return newState.incrementLastEventIndex()
  }

  private fun incrementLastEventIndex() =
    copy(lastAppliedEventIndex = lastAppliedEventIndex + 1)

  private fun apply(event: MeetupEventCapacityIncreased) =
    copy(capacity = event.newCapacity)

  private fun apply(event: UserSubscribedToMeetupEvent): MeetupEventState {
    val subscription = Subscription(event.userId, event.registrationTime, false)
    val newSubscriptions = subscriptions.add(subscription)
    return copy(subscriptions = newSubscriptions)
  }

  private fun apply(event: UserAddedToMeetupEventWaitingList): MeetupEventState {
    val subscription = Subscription(event.userId, event.registrationTime, true)
    val newSubscriptions = subscriptions.add(subscription)
    return copy(subscriptions = newSubscriptions)
  }

  private fun apply(event: UserCancelledMeetupSubscription): MeetupEventState {
    val (newSubscriptions) = subscriptions.removeBy(event.userId)
    return copy(subscriptions = newSubscriptions)
  }

  private fun apply(event: UserMovedFromWaitingListToParticipants): MeetupEventState {
    val userSubscription = subscriptions.findBy(event.userId)
    val newSubscriptions =
      if (userSubscription != null) subscriptions.confirm(userSubscription) else subscriptions
    return copy(subscriptions = newSubscriptions)
  }

  private fun apply(event: UsersMovedFromWaitingListToParticipants): MeetupEventState {
    val subscriptions = event.userIdList.mapNotNull { subscriptions.findBy(it) }
    val newSubscriptions = this.subscriptions.confirm(subscriptions)
    return copy(subscriptions = newSubscriptions)
  }
}

fun projectStateFrom(
  events: List<MeetupBaseEvent>,
  snapshot: MeetupEventState = MeetupEventState(0, 0, "", LocalDateTime.MIN),
): MeetupEventState =
  events.drop(snapshot.lastAppliedEventIndex + 1)
    .fold(snapshot) { state, event ->
      state.applyEvent(event)
    }

fun newMeetup(
  id: Long,
  eventName: String,
  eventCapacity: Int,
  startTime: LocalDateTime
): MeetupEvent {
  val meetupEventRegistered = MeetupEventRegistered(id, eventName, eventCapacity, startTime)
  return MeetupEvent(listOf(meetupEventRegistered))
}

data class MeetupEvent(
  val state: MeetupEventState,
  val events: List<MeetupBaseEvent>,
  val version: Int = -1,
) {

  constructor(events: List<MeetupBaseEvent>) : this(projectStateFrom(events), events)
  constructor(events: List<MeetupBaseEvent>, version: Int) : this(projectStateFrom(events), events, version)

  fun subscribe(userId: String, registrationTime: Instant): MeetupEvent {
    val event = if (state.isFull) {
      UserAddedToMeetupEventWaitingList(state.id, userId, registrationTime)
    } else {
      UserSubscribedToMeetupEvent(state.id, userId, registrationTime)
    }
    return this.apply(event)
  }

  fun unsubscribe(userId: String, meetupEventId: Long): MeetupEvent {

    val event1 = state.subscriptions.subscriptionOf(userId)
      .map { UserCancelledMeetupSubscription(meetupEventId, userId) }

    val event2 = state.subscriptions.subscriptionOf(userId)
      .filter { !it.isInWaitingList }
      .flatMap { state.subscriptions.firstInWaitingList }
      .map { firstInWaitingList -> UserMovedFromWaitingListToParticipants(
        meetupEventId,
        firstInWaitingList.userId,
        event1.get()
      ) }

   return this.apply(listOf(event1, event2))
  }

  fun increaseCapacityTo(newCapacity: Int): MeetupEvent {
    val meetupCapacityIncreased = MeetupEventCapacityIncreased(state.id, newCapacity)

    val movedUsers = UsersMovedFromWaitingListToParticipants(
      state.id,
      state.subscriptions.firstInWaitingList(newCapacity - state.capacity).map { it.userId },
      meetupCapacityIncreased
    )

    return this.apply(listOf(meetupCapacityIncreased, movedUsers))
  }

  private fun apply(event: MeetupBaseEvent) = apply(listOf(event))

  private fun apply(newEvents: List<MeetupBaseEvent>) = MeetupEvent(
    events = this.events + newEvents,
    state = projectStateFrom(this.events + newEvents, state),
    version = version,
  )
}

private fun listOf(
  event1: Optional<UserCancelledMeetupSubscription>,
  event2: Optional<UserMovedFromWaitingListToParticipants>
): List<MeetupBaseEvent> = Stream.concat(event1.stream(), event2.stream())
  .toList()
